db/sqldb/user.go

```diff
@@ -52,6 +52,43 @@ func (d *DB) UserByAuthn(tx db.Tx, authnMechanism pacta.AuthnMechanism, authnID
 	return exactlyOne("user", fmt.Sprintf("%s:%s", authnMechanism, authnID), us)
 }
 
+func (d *DB) GetOrCreateUserByAuthn(tx db.Tx, authnMechanism pacta.AuthnMechanism, authnID string, email string) (*pacta.User, error) {
+	var user *pacta.User
+	err := d.RunOrContinueTransaction(tx, func(tx db.Tx) error {
+		u, err := d.userByAuthn(tx, authnMechanism, authnID)
+		if err == nil {
+			user = u
+			return nil
+		}
+		if !db.IsNotFound(err) {
+			return fmt.Errorf("looking up user by authn: %w", err)
+		}
+		canonizedEmail, err := pacta.CanonizeEmail(email)
+		if err != nil {
+			return fmt.Errorf("canonizing email: %w", err)
+		}
+		uID, err := d.CreateUser(tx, &pacta.User{
```

Comment:
As far as I can tell, this function is only used internally and should probably not be exported.

Similar to above comment, we should really have exactly one public function that can create users - it's a somewhat complicated operation and we don't want multiple places doing it differently.

Reply:
agreed. fixed.